File: app/src/main/java/com/example/newsapp/adapters/SourcesAdapter.java
package com.example.newsapp.adapters;

import android.content.Context;
import android.os.Bundle;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.example.newsapp.Fragments.MainFragment;
import com.example.newsapp.R;

import java.util.List;

import data.datamodels.Sources;

public class SourcesAdapter extends
        RecyclerView.Adapter<SourcesAdapter.SourceViewHolder> {

    private List<Sources> mSourceDataSet;
    private LayoutInflater layoutInflater;
    private Context context;

    public SourcesAdapter(Context context, List<Sources> sourceList){
        mSourceDataSet = sourceList;

        this.layoutInflater = LayoutInflater.from(context);
        this.context = context;

    }

    @Override
    public SourceViewHolder onCreateViewHolder(ViewGroup viewGroup,
                                                int viewType){
        View view = layoutInflater
                .inflate(R.layout.card_source_data,viewGroup, false);
        SourceViewHolder sourceViewHolder = new SourceViewHolder(view);

        return sourceViewHolder;

    }

    @Override
    public void onBindViewHolder(SourceViewHolder holder, int position){
        Sources currentSource = mSourceDataSet.get(position);
        holder.setData(currentSource,position);


    }

    @Override
    public int getItemCount() {
        return mSourceDataSet.size();
    }

    class SourceViewHolder extends RecyclerView.ViewHolder {

        TextView tv_name, tv_description, tv_category, tv_language, tv_country;

        public SourceViewHolder(View sourcesView){
            super(sourcesView);
            tv_name = sourcesView.findViewById(R.id.tv_name);
            tv_description = sourcesView.findViewById(R.id.tv_description);
            tv_category = sourcesView.findViewById(R.id.tv_category);
            tv_language = sourcesView.findViewById(R.id.tv_language);
            tv_country = sourcesView.findViewById(R.id.tv_country);

        }

        public void setData(final Sources current, final int position){


            this.tv_name.setText(current.getName());
            this.tv_description.setText(current.getDescription());
             FragmentManager fragmentManager = ((AppCompatActivity)context)
                    .getSupportFragmentManager();

            tv_description.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {

                    int fragmentContainer;
                    FragmentTransaction fragmentTransaction;

                    MainFragment mainFragment = new MainFragment();
                    Bundle bundle = new Bundle();
                    bundle.putString("DataFlag","data_by_url");
                    bundle.putString("DataUrl", current.getUrl());

                    mainFragment.setArguments(bundle);
                    fragmentContainer = R.id.fr_main_holder;
                    fragmentTransaction = fragmentManager.beginTransaction();

                    fragmentTransaction.replace(fragmentContainer, mainFragment);
                    fragmentTransaction.addToBackStack(null);
                    fragmentTransaction.commitAllowingStateLoss();

                }
            });
            this.tv_category.setText(current.getCategory().substring(0,1).toUpperCase()
                    .concat(current.getCategory().substring(1)));
            this.tv_language.setText(current.getLanguage());
            this.tv_country.setText(current.getCountry().toUpperCase());


        }

    }
}

File: app/src/main/java/com/example/newsapp/adapters/SourcesAdapter.kt
package com.example.newsapp.adapters

import android.content.Context
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.example.newsapp.Fragments.MainFragment
import com.example.newsapp.R
import com.example.newsapp.data.datamodels.Sources

class SourcesAdapter(context: Context, sourceList: List<Sources>) :
        RecyclerView.Adapter<SourcesAdapter.SourceViewHolder>() {


    var mSourceList = sourceList
    var layoutInflater = LayoutInflater.from(context)


    override fun onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): SourceViewHolder {

        var view: View = layoutInflater.inflate(R.layout.card_source_data, viewGroup, false)
        return SourceViewHolder(view)

    }

    override fun onBindViewHolder(sourceViewHolder: SourceViewHolder, position: Int) {

        var currentSource: Sources = mSourceList.get(position)
        sourceViewHolder.setData(currentSource, position)
    }

    override fun getItemCount(): Int {
        return mSourceList.size

    }

    class SourceViewHolder(var view: View) : RecyclerView.ViewHolder(view) {

        var tv_name: TextView = view.findViewById(R.id.tv_name)
        var tv_description: TextView = view.findViewById(R.id.tv_description)
        var tv_category: TextView = view.findViewById(R.id.tv_category)
        var tv_language: TextView = view.findViewById(R.id.tv_language)
        var tv_country: TextView = view.findViewById(R.id.tv_country)
        var mContext = view.context

        fun setData(current: Sources, position: Int) {
            tv_name.setText(current.name)
            tv_description.setText(current.description)
            var fragmentManager = (mContext as AppCompatActivity).supportFragmentManager

            var mainFragment = MainFragment()
            var bundle = Bundle()
            bundle.putString(mContext.getString(R.string.data_by_url_flag_key),
                    mContext.getString(R.string.data_by_url_value))
            bundle.putString(mContext.getString(R.string.current_url_key), current.url)

            mainFragment.arguments = bundle

            tv_description.setOnClickListener {
                var fragmentContainer: Int = R.id.fr_main_holder
                var fragmentTransaction = fragmentManager.beginTransaction()

                fragmentTransaction.replace(fragmentContainer, mainFragment)
                        .addToBackStack(null)
                        .commitAllowingStateLoss()

            }

            tv_category.text = (current.category.substring(0, 1).toUpperCase()
                    + current.category.substring(1))
            tv_language.setText(current.language)
            this.tv_country.setText(current.country.toUpperCase())

        }


    }


}